apps/designer/app/routes/designer/$id.tsx

```diff
@@ -1,30 +1,82 @@
 import { useLoaderData } from "@remix-run/react";
-import { LoaderFunction } from "@remix-run/node";
-import type { Project } from "@webstudio-is/react-sdk";
+import path from "path";
+import {
+  ActionFunction,
+  LoaderFunction,
+  unstable_createFileUploadHandler,
+  unstable_parseMultipartFormData,
+} from "@remix-run/node";
+import type { Project, Asset } from "@webstudio-is/react-sdk";
 import { Designer, links } from "~/designer";
 import * as db from "~/shared/db";
 import config from "~/config";
 import env from "~/env.server";
+import { z } from "zod";
 export { links };
 
 export const loader: LoaderFunction = async ({ params }) => {
   if (params.id === undefined) throw new Error("Project id undefined");
   const project = await db.project.loadById(params.id);
+  const assets = await db.assets.loadByProject(params.id);
   if (project === null) {
     return { errors: `Project "${params.id}" not found` };
   }
-  return { config, project, env };
+  return { config, assets, project, env };
 };
 
 type Data = {
   config: typeof config;
   project: Project;
+  assets: Asset[];
 };
 
 type Error = {
   errors: "string";
 };
 
+const ImageUpload = z.object({
+  name: z.string(),
+  type: z.string(),
+});
+type ImageUpload = z.infer<typeof ImageUpload>;
+
+export const action: ActionFunction = async ({ request, params }) => {
+  const uploads = path.join(__dirname, "../public");
```

Comment:
we shouldn't use __dirname (logic where to upload shouldn't depend on where this code is, if we mode code it breaks), also I think this is still the wrong path to public

Reply:
It's not, it works

But what should I use?

Reply:
is there anything in env that we can use? kinda need to have a stable reference to a root somewhere

Reply:
Messaed this on discord bt this dirname is the build and thats stable